### PR TITLE
Migrate Mirage 2 build from bower and Ruby sass to npm and node-sass

### PR DESCRIPTION
**Pull request by alanorth**
## References
* [Recent mailing list post about Mirage 2 build issues related to torquebox.org](https://groups.google.com/g/dspace-tech/c/7_fUBxsxkhs/m/xTpRzyDUAQAJ)

## Description
In short, this pull request replaces bower and Ruby sass with npm and node-sass in order to fix the broken Mirage 2 build when using the "slow" maven build. This replaces the [previous pull request that was using yarn](https://github.com/DSpace/DSpace/pull/8288).

For more background, the Mirage 2 build currently uses bower for package management and Ruby sass/compass for Sass processing. Bower has been deprecated for years and is an increasing source of obscure build issues. Compass is Ruby based and requires the user to either preinstall Ruby and several gems or use those provided by JRuby via torquebox.org during build time. As of last week the torquebox.org domain expired and is likely abandoned, making it impossible to build Mirage 2.

## Instructions for Reviewers
Test building Mirage 2 at your site using the `-Dmirage2.on=true -Dmirage2.deps.included=true` option to maven. I have tested both vanilla Mirage 2 and a slightly modified one with style overrides in `dspace/modules/xmlui-mirage2`.

## Notes
List of changes in this PR:
*  Allow building Mirage 2 with Node.js 12 and use eirslett/frontend-maven-plugin to install Node.js 12 during maven "slow" build mode. JavaScript packages that were previously installed by bower are now installed by npm.
* Update CI to use latest Node.js 12 and remove useless pre-requisite step (as GitHub CI already has npm and grunt installed).
* Remove JRuby and compass stuff from build and use node-sass as the Sass processor and essentially polyfill the compass Sass mixins using compass-mixins from npmjs.com.
* Use a static copy of Modernizr.min.js because the version we need is not available on npmjs.com.
* Update grunt build tools

## TODO
Once this is merged we will also need to update the DSpace installation documentation on the wiki.

### CODE REVIEW

File: dspace-xmlui-mirage2/pom.xml [hunk 1]
```diff
@@ -53,6 +53,7 @@
                         <exclude>**/readme.txt</exclude>
                         <exclude>**/.bowerrc</exclude>
```
Comment (mwoodiupui) [inline dspace-xmlui-mirage2/pom.xml:54]: Do we still need `.bowerrc`?  Maybe remove here and in `dspace-xmlui-mirage2/src/main/webapp`?

Comment (alanorth) [reply dspace-xmlui-mirage2/pom.xml:54]: Good catch! I've removed them.

File: dspace/modules/xmlui-mirage2/src/main/webapp/themes/Mirage2/styles/_style.scss [hunk 2]
```diff
@@ -58,4 +58,8 @@
   #aspect_submission_StepTransformer_item_publication-title{
     margin-bottom:5px;
   }
-}
\ No newline at end of file
+}
+
+body {
+    color: green;
```
Comment (kshepherd) [inline dspace/modules/xmlui-mirage2/src/main/webapp/themes/Mirage2/styles/_style.scss:64]: This file looks like an unintended inclusion from a local theme?

**Review by kshepherd — COMMENTED** [review]
Looks great to me, thakns @alanorth  
I've tested a build locally and it builds quickly and successfully, and Mirage 2 is working
I left a comment about a "color:green" added to body text in the Mirage 2 _style.scss overlay which I wasn't quite sure about.

Once that little bit is cleared up I'm good to approve!

File: dspace/modules/xmlui-mirage2/src/main/webapp/themes/Mirage2/styles/_style.scss (hunk 2 above)
Comment (alanorth) [reply dspace/modules/xmlui-mirage2/src/main/webapp/themes/Mirage2/styles/_style.scss:64]: Good catch, yes that was added accidentally with `git add -u`. Removed!

**Review by kshepherd — APPROVED** [review]
+1 reviewed and tested

**Comment by kshepherd** [comment]
I was just trying this out with a heavily-customised repo I help with and it made me think we should include some notes in the documentation or release notes about how some of these changes work, so people with custom themes that might be importing extra node deps, including other scss with references to the old paths to deps, etc, can update their local themes to match the new way dependencies are managed

**Comment by johnpinto1** [comment]
A patch for the 21 commits fixed the build for Edinburgh Datashare. Thanks @alanorth. I hope it will be merged into the Dspace 6 branch Dspace: dspace_6.x soon.

**Comment by alanorth** [comment]
Thank you, @johnpinto1. I suspect you had to move some dependencies over from `.bowerrc` to `package.json`, update some include paths from `vendor` to `node_modules`, and some include paths in SCSS? I tried to capture some of those tips in the [updated Mirage 2 readme.md](https://github.com/DSpace/DSpace/pull/8292/commits/7bc99ec88a4240349969272de35b81dd207148e6) included in this PR.

I think we are close to merging this now, thanks to your +1. Also, I have tested locally with Node.js 14 and it works, so I will push to get that in here too since Node.js 12 is no longer supported upstream.

**Review by tdonohue — APPROVED** [review]
👍 The code here all looks reasonable to me.  I haven't been able to test it myself, but I trust the tests of others (and it sounds like at least 3 others have tested it successfully)

**Comment by tdonohue** [comment]
NOTE: This issue was also fixed on `dspace-5_x`, but in a different manner. In that branch, I simply found a different way to pull in the required RubyGems, see this commit which provided the fix: https://github.com/DSpace/DSpace/commit/0428da8dab7a02c592ec96ab9cea545f5b6de42d

**Pull request merged**